lib/grunt/deprecate.js

```diff
@@ -0,0 +1,116 @@
+var grunt = require('../grunt');
+
+function deprecate(obj, property, message) {
+  if (Array.isArray(obj)) {
+    obj.forEach(function(item) {
+      deprecate(item.obj, item.property, item.message);
+    });
+    return;
+  }
+  var logged = false;
+  function warn() {
+    var hideDeprecation = grunt.option('hide-deprecations');
+    if (!hideDeprecation && !logged) {
+      if (grunt.option('stack')) {
+        grunt.log.warn(Error(message).stack);
+      } else {
+        grunt.log.warn(message);
+      }
+      logged = true;
+    }
+  }
+  var orig = obj[property];
+  Object.defineProperty(obj, property, {
+    enumerable: true,
+    configurable: true,
+    set: function(val) {
+      warn();
+      orig = val;
+    },
+    get: function() {
+      warn();
+      return orig;
+    }
+  });
+}
+module.exports = deprecate;
+
+deprecate([
```

Comment:
It might make sense to call `deprecate` in a different file than where it's defined... but perhaps I'm overthinking things!

Reply:
Separating the API from the apply makes sense to me. I moved the deprecations into their own file called `deprecations.js`.